### PR TITLE
chore: Build Wasmer on musl

### LINKED ISSUE #1482: Compatibility with Alpine Linux
Thanks for proposing a new feature!

### Motivation

Many tiny Docker deployments use Alpine Linux as its base to reduce size of images and containers. Currently, wasmer is incompatible with such environments because it's compiled against glibc.

### Proposed solution

Provide a build of wasmer that can work with musl libc as well.

### Additional context

For background, initially I stumbled upon this due to non-sensical error after installation:

```bash
/ # ls /root/.wasmer/bin/wasmer
/root/.wasmer/bin/wasmer
/ # /root/.wasmer/bin/wasmer
sh: /root/.wasmer/bin/wasmer: not found
```

After searching around, I found explanation of a similar issue as well a workaround that creates a symbolic link for musl libc pretending to be full one: https://stackoverflow.com/questions/34729748/installed-go-binary-not-found-in-path-on-alpine-linux-docker

That helped a bit, but there are still missing functions due to differences in libc implementations:

```bash
/ # mkdir /lib64
/ # ln -s /lib/libc.musl-x86_64.so.1 /lib64/ld-linux-x86-64.so.2
/ # /root/.wasmer/bin/wasmer
Error relocating /root/.wasmer/bin/wasmer: __snprintf_chk: symbol not found
Error relocating /root/.wasmer/bin/wasmer: mallinfo: symbol not found
Error relocating /root/.wasmer/bin/wasmer: __fprintf_chk: symbol not found
Error relocating /root/.wasmer/bin/wasmer: __vsnprintf_chk: symbol not found
Error relocating /root/.wasmer/bin/wasmer: __longjmp_chk: symbol not found
```

Hence, I think it would be best if wasmer would be built against musl as well.

### LINKED ISSUE #1766: alpine build error
### Describe the bug

try build from alpine, get this error:

```sh

  --- stderr
  autoreconf: Entering directory `.'
  autoreconf: configure.ac: not using Gettext
  autoreconf: running: aclocal -I m4
  autoreconf: configure.ac: tracing
  autoreconf: running: libtoolize --copy
  autoreconf: running: /usr/bin/autoconf
  autoreconf: running: /usr/bin/autoheader
  autoreconf: running: automake --add-missing --copy --no-force
  configure.ac:31: installing './compile'
  configure.ac:19: installing './install-sh'
  configure.ac:19: installing './missing'
  Makefile.am: installing './depcomp'
  doc/Makefile.am:3: installing 'doc/mdate-sh'
  doc/Makefile.am:3: installing 'doc/texinfo.tex'
  autoreconf: Leaving directory `.'
  ../src/java_raw_api.c: In function 'ffi_java_raw_call':
  ../src/java_raw_api.c:317:3: warning: 'ffi_java_raw_to_ptrarray' is deprecated [-Wdeprecated-declarations]
    317 |   ffi_java_raw_to_ptrarray (cif, raw, avalue);
        |   ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c:76:1: note: declared here
     76 | ffi_java_raw_to_ptrarray (ffi_cif *cif, ffi_java_raw *raw, void **args)
        | ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c: In function 'ffi_java_translate_args':
  ../src/java_raw_api.c:328:3: warning: 'ffi_java_raw_size' is deprecated [-Wdeprecated-declarations]
    328 |   ffi_java_raw *raw = (ffi_java_raw*)alloca (ffi_java_raw_size (cif));
        |   ^~~~~~~~~~~~
  ../src/java_raw_api.c:45:1: note: declared here
     45 | ffi_java_raw_size (ffi_cif *cif)
        | ^~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c:331:3: warning: 'ffi_java_ptrarray_to_raw' is deprecated [-Wdeprecated-declarations]
    331 |   ffi_java_ptrarray_to_raw (cif, avalue, raw);
        |   ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c:157:1: note: declared here
    157 | ffi_java_ptrarray_to_raw (ffi_cif *cif, void **args, ffi_java_raw *raw)
        | ^~~~~~~~~~~~~~~~~~~~~~~~
  ../src/java_raw_api.c: In function 'ffi_prep_java_raw_closure':
  ../src/java_raw_api.c:369:3: warning: 'ffi_prep_java_raw_closure_loc' is deprecated [-Wdeprecated-declarations]
    369 |   return ffi_prep_java_raw_closure_loc (cl, cif, fun, user_data, cl);
        |   ^~~~~~
  ../src/java_raw_api.c:337:1: note: declared here
    337 | ffi_prep_java_raw_closure_loc (ffi_java_raw_closure* cl,
        | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  ar: `u' modifier ignored since `D' is the default (see `U')
  ar: `u' modifier ignored since `D' is the default (see `U')
  thread 'main' panicked at 'Unable to find libclang: "the `libclang` shared library at /usr/lib/libclang.so.10 could not be opened: Dynamic loading not supported"', /opt/build/php/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.52.0/src/lib.rs:1895:31
  stack backtrace:
     0: rust_begin_unwind
               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/panicking.rs:483:5
     1: core::panicking::panic_fmt
               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/core/src/panicking.rs:85:14
     2: core::option::expect_none_failed
               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/core/src/option.rs:1268:5
     3: core::result::Result<T,E>::expect
     4: core::ops::function::FnOnce::call_once
     5: lazy_static::lazy::Lazy<T>::get::{{closure}}
     6: std::sync::once::Once::call_once::{{closure}}
     7: std::sync::once::Once::call_inner
               at /rustc/ffa2e7ae8fbf9badc035740db949b9dae271c29f/library/std/src/sync/once.rs:419:21
     8: std::sync::once::Once::call_once
     9: <bindgen::ensure_libclang_is_loaded::LIBCLANG as core::ops::deref::Deref>::deref
    10: bindgen::ensure_libclang_is_loaded
    11: bindgen::Bindings::generate
    12: bindgen::Builder::generate
    13: build_script_build::generate_bindings
    14: build_script_build::main
    15: core::ops::function::FnOnce::call_once
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
warning: build failed, waiting for other jobs to finish...
warning: 2 warnings emitted

error: build failed
```

### Steps to reproduce

build from alpine/edge,  already install musl-dev clang-dev llvm-dev. 

```sh
ls -alFh /usr/lib/libclang.so.10
-rwxr-xr-x    1 root     root       26.0M Jul 30 14:48 /usr/lib/libclang.so.10*
 ldd /usr/lib/libclang.so.10
	/lib/ld-musl-x86_64.so.1 (0x7fcca8fcf000)
	libLLVM-10.so => /usr/lib/../lib/libLLVM-10.so (0x7fcca38de000)
	libstdc++.so.6 => /usr/lib/../lib/libstdc++.so.6 (0x7fcca373d000)
	libgcc_s.so.1 => /usr/lib/../lib/libgcc_s.so.1 (0x7fcca3723000)
	libc.musl-x86_64.so.1 => /lib/ld-musl-x86_64.so.1 (0x7fcca8fcf000)
	libffi.so.7 => /usr/lib/../lib/libffi.so.7 (0x7fcca3718000)
	libz.so.1 => /lib/libz.so.1 (0x7fcca36fe000)
	libxml2.so.2 => /usr/lib/../lib/libxml2.so.2 (0x7fcca35d4000)
	liblzma.so.5 => /usr/lib/../lib/liblzma.so.5 (0x7fcca35b1000)
```

### Expected behavior

expect build c-api static lib.

### PR DESCRIPTION
**Pull request by jubianchi**
This patch adds a specific build for musl. Currently, it will only support JIT engine.

I also changes the workflow definition a bit so we don't depend on the OS name but rather on the environment ID which is our convention.

Closes #1482
Closes #1766

# Description

# Review

- [ ] Add a short description of the the change to the CHANGELOG.md file

### CODE REVIEW

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by syrusakbary** [comment]
Why no support for the native engine?

**Review by Neves-Adriano — APPROVED** [review]
👍🏻

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
@syrusakbary there are problem with the native engine and linkage: currently, musl is statically linked and it crates problem with dynamic loading.

I chose to disable these things for now and then experiment more on using musl as a shared library. 

Reading the rust documentation it seems as easy as using a `target-feature=-crt-static` but it does not work (and I don't know why yet)

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

File: Makefile
Comment (Hywan) [inline Makefile:93]: :+1:

File: Makefile
Comment (Hywan) [inline Makefile:392]: I presume you wanted to write:

```suggestion
	if [ -f "target/$(HOST_TARGET)/wasmer-headless" ]; then \
```

But why moving `libwasmer_c_api.so` to `libwasmer.so` if `wasmer-headless` exists?

File: examples/engine_cross_compilation.rs
Comment (Hywan) [inline examples/engine_cross_compilation.rs:109]: I assumed you wanted to remove the `#[cfg]` but keep the `fn test_cross_compilation`, is that correct?

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [inline .github/workflows/main.yaml:188]: I wonder if it's possible to have this be one step by using the matrix to set the env vars ahead of time...

That would be optimal but I find it easy to believe that that's something that sounds way easier than it actually is to do.

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [inline .github/workflows/main.yaml:192]: At some point we may just want to encode the logic in the matrix directly as a variable like `should-test-capi` and have `true`/`false` for each target.

I'll leave it up to your judgement what you think is better, just thinking out loud

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [inline .github/workflows/main.yaml:208]: Yeah, if we could set env vars then we wouldn't need to duplicate this either

Comment (MarkMcCaskey) [reply .github/workflows/main.yaml:208]: It'll also scale better as we add more things

File: Makefile
Comment (MarkMcCaskey) [reply Makefile:392]: I'm also unclear on the logic here

**Review by MarkMcCaskey — COMMENTED** [review]
Looks pretty good to me

One thing to consider before shipping: does `wapm` build properly on musl?

It might also be a good idea to get this into `wasmer-nightly`, some recent changes I mad ein `main.yaml` also need to be ported to wasmer-nightly I think... 🤔

File: .github/workflows/main.yaml
Comment (jubianchi) [reply .github/workflows/main.yaml:192]: @MarkMcCaskey I totally agree.

I started my PR with much more things in the workflow/makefile (improvements, optimizations) but ran into problem and reverted everything to try to find the root cause.

I think we can explore this in a dedicated PR.

**Comment by jubianchi** [comment]
bors try

File: .github/workflows/main.yaml
Comment (Hywan) [inline .github/workflows/main.yaml:184]: For the sake of consistency:

```suggestion
        if: matrix.build != 'macos-arm64' && matrix.build != 'linux-musl-x64'
```

File: .github/workflows/main.yaml
Comment (Hywan) [reply .github/workflows/main.yaml:192]: I don't understand. We don't test the C API with musl, but we build it (just below this line)? Would be nice to test it, don't you think :-p?

**Comment by jubianchi** [comment]
> I don't understand. We don't test the C API with musl, but we build it (just below this line)? Would be nice to test it, don't you think :-p?

Some tests are failing with musl but I wanted to have the library so I can use it to test without having to build it manually.

I will fix the workflow to not build the C API with musl: it does not really work so let's not ship it.

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

**Comment by jubianchi** [comment]
bors trybors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try-

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

**Comment by jubianchi** [comment]
bors try

File: .cargo/config.toml
Comment (nlewycky) [inline .cargo/config.toml:3]: (No action required.) Could you explain?

File: .github/workflows/main.yaml
Comment (nlewycky) [inline .github/workflows/main.yaml:223]: Why would there ever be a libwasmer without a `.` in the name? Maybe this could have a comment?

File: .github/workflows/main.yaml
Comment (nlewycky) [inline .github/workflows/main.yaml:336]: Any particular reason not to use `make test test-capi`?

File: Makefile
Comment (nlewycky) [inline Makefile:330]: How about TARGET_DIR := target/$(TARGET)/release when true or TARGET_DIR := target/release when TARGET is blank? Then we can simplify a lot of paths.

My underlying problem with this is that a single variable with one name switched its "logical type" from being a string containing just the target type, to being a directory name.

File: lib/c-api/tests/.gitignore
Comment (nlewycky) [inline lib/c-api/tests/.gitignore:4]: Is this change related to musl?

File: .cargo/config.toml
Comment (jubianchi) [reply .cargo/config.toml:3]: Something I forgot to remove while testing different configurations but I like having trailing-comma when it's allowed to reduce diff when adding lines into lists.

File: .github/workflows/main.yaml
Comment (jubianchi) [reply .github/workflows/main.yaml:336]: only reason is to keep the script as close as possible to the workflow steps we use in the previous job.

I'll make the tests steps a one-liner ;)

File: lib/c-api/tests/.gitignore
Comment (jubianchi) [reply lib/c-api/tests/.gitignore:4]: nope.

It can be extracted to a dedicated PR if needed.

File: .github/workflows/main.yaml
Comment (jubianchi) [reply .github/workflows/main.yaml:223]: This step will be removed:  the `TARGET` variable in the Makefile will do the job :)

**Comment by jubianchi** [comment]
bors try

File: .github/workflows/main.yaml
Comment (Hywan) [inline .github/workflows/main.yaml:444]: It seems like this code is a duplication of the code below (see my comment).

File: .github/workflows/main.yaml
Comment (Hywan) [inline .github/workflows/main.yaml:480]: This is the duplicated code I'm mentionning in my comment above (line 435).

We should keep this block, and remove the previous one. The `id` key isn't necessary.

File: Makefile
Comment (Hywan) [inline Makefile:654]: I like this new approach. It's more “error-proof”.

**Review by Hywan — APPROVED** [review]
The asset upload duplication must be fixed, but apart from that, I'm approving this PR. Good job!

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [inline .github/workflows/main.yaml:41]: Huh, I had just assumed that line was always required

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [inline .github/workflows/main.yaml:175]: Are target and build the same here? It seems like they could be different

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [inline .github/workflows/main.yaml:229]: This looks like a work-around for aarch64 macos, do we not need it anymore?

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [inline .github/workflows/main.yaml:300]: is the `:/` intentional or should this just be `/`?

File: Makefile
Comment (MarkMcCaskey) [reply Makefile:654]: I was actually just about to comment that I think it's more error prone as we'll silently fail if we're missing any of these files and will just have to wait until a user reports or we notice that we didn't release one of the files we intended to.

Ideally our packaging logic would be tested and not live in our CI file. We've lost a ton of things done in CI over time from all the CI migrations we've done -- I wouldn't put anything too critical there

Comment (MarkMcCaskey) [reply Makefile:654]: I do think the above code is better though, but we do need tests for our packaging logic and the change above makes the need for tests greater.

File: .github/workflows/main.yaml
Comment (jubianchi) [reply .github/workflows/main.yaml:300]: It intentional, this part is a Docker volume mapping which syntax is `<path-on-the-host>:<path-in-the-container>`

File: .github/workflows/main.yaml
Comment (jubianchi) [reply .github/workflows/main.yaml:229]: With the new `TARGET_DIR` variable in the `Makefile` this step is not required anymore.

File: .github/workflows/main.yaml
Comment (jubianchi) [reply .github/workflows/main.yaml:175]: `matrix.target` and `matrix.os` are names defined by either Rust or GitHub. `matrix.build` is something we chosse on our side so I prefer to rely on our naming convention here. 

If GitHub forces us to change the `os` this will not impact the workflow. Same for `target` (even if it's unlikely they will change)

File: Makefile
Comment (jubianchi) [reply Makefile:654]: @MarkMcCaskey the problem with our previous strategy here was we relied on the host OS to define the files to copy.

This is not always the right thing to do, for example when we cross-compile this is why I changed that. Also, our strategy was not adaptative: if a user decides to change his build configuration (for example move from `cdylib` to `staticlib`) he'll have to change the Makefile. With this change, everything will automatically work.

File: .github/workflows/main.yaml
Comment (MarkMcCaskey) [reply .github/workflows/main.yaml:175]: Sure that makes sense, I just want to check that there aren't extra semantics here. If I remember correctly, arm64 Mac used `target` or some other key for a very specific purpose and I think the difference might have been important because of cross-compilation or something

File: Makefile
Comment (MarkMcCaskey) [reply Makefile:654]: Yeah that makes sense, my one concern is just that it's less specific. Rather than encoding an expected state of the world "given condition X, we must have files Y, Z, ..." we just take whatever is there. So any bug that affects the files we produce will now go unnoticed.

I think the real fix is to just have actual testing for our release process but I think that's out of scope for this PR.

File: .github/workflows/main.yaml
Comment (jubianchi) [reply .github/workflows/main.yaml:175]: @markandrus yeah, to build for M1 we cross compile so we define (in the matrix) the Rust target to use and we were also using this in our conditionals.

But now, we have a problem: when GitHub will support M1 directly, we won't cross-compile anymore (we won't force the target through the matrix). Without my change, this would also require to change all the conditionals. With my change, we only update the matrix and we are done.

File: Makefile
Comment (Hywan) [reply Makefile:657]: Let's accept the following patch, and then immediately open a new PR to test the `package/` directory? We also have various scheduled tests that run based`wasmer-nightly` (like in `wasmer-go`, `wasmer-php` etc.).

**Pull request merged**
